File: clients/Go/model/action.go
package model

import "io"
import . "aicup2020/stream"

type Action struct {
    EntityActions map[int32]EntityAction
}
func NewAction(entityActions map[int32]EntityAction) Action {
    return Action {
        EntityActions: entityActions,
    }
}
func ReadAction(reader io.Reader) Action {
    result := Action {}
    EntityActionsSize := ReadInt32(reader)
    result.EntityActions = make(map[int32]EntityAction)
    for i := int32(0); i < EntityActionsSize; i++ {
        var EntityActionsKey int32
        EntityActionsKey = ReadInt32(reader)
        var EntityActionsValue EntityAction
        EntityActionsValue = ReadEntityAction(reader)
        result.EntityActions[EntityActionsKey] = EntityActionsValue
    }
    return result
}
func (value Action) Write(writer io.Writer) {
    WriteInt32(writer, int32(len(value.EntityActions)))
    for EntityActionsKey, EntityActionsValue := range value.EntityActions {
        WriteInt32(writer, EntityActionsKey)
        EntityActionsValue.Write(writer)
    }
}

File: clients/Python-PyPy/model/action.py
from .entity_action import EntityAction
class Action:
    def __init__(self, entity_actions):
        self.entity_actions = entity_actions
    @staticmethod
    def read_from(stream):
        entity_actions = {}
        for _ in range(stream.read_int()):
            entity_actions_key = stream.read_int()
            entity_actions_value = EntityAction.read_from(stream)
            entity_actions[entity_actions_key] = entity_actions_value
        return Action(entity_actions)
    def write_to(self, stream):
        stream.write_int(len(self.entity_actions))
        for key, value in self.entity_actions.items():
            stream.write_int(key)
            value.write_to(stream)
    def __repr__(self):
        return "Action(" + \
            repr(self.entity_actions) + \
            ")"
